Анализ производительности приложения

У нас есть сгенерированные логи сервера, по этим данным мы проведем анализ сервера, а то есть: 
1. сколько всего запросов в файле;
2. сколько было ошибок;
3. какой endpoint вызывался чаще всего;
4. среднее время ответа сервера.

In [1]:
with open("server_logs.txt", "r") as file:
    lines = file.readlines()  # читает все строки в список
print(f"Количество строк: {len(lines)}")

Количество строк: 50


Так мы сможем просмотреть все запросы которые есть в файле и получить число запросов, котоыре записаны в файле

Заранее знаем, что в файле есть дата, время, метод, endpoint, статус кода и время ответа

In [2]:
for line in lines:
    parts = line.strip().split()
    date = parts[0]  # разделяет строку на части и берет первую часть (дату)
    time = parts[1]  # берет вторую часть (время) 
    method = parts[2]  # берет третью часть (метод)
    endpoint = parts[3]  # берет четвертую часть (endpoint)
    status_code = parts[4]  # берет пятую часть (код статуса)
    time_to_answear = parts[5]  # берет шестую часть (время ответа) 

После, того как мы разделили все логи по частям, мы можем при помощи счетчика считать количество ошибок со статусом 404 и 500, каждый раз если parts[4] начинается с 4 или 5, то счетчик будет прибавлять 1, им в терминал выведет отдльной стракой количество ошибок.

In [3]:
errors = 0
with open("server_logs.txt", "r") as file:
    lines = file.readlines()  # читает все строки в список
print(f"Количество строк: {len(lines)}")
for line in lines:
    parts = line.strip().split()
    date = parts[0]  # разделяет строку на части и берет первую часть (дату)
    time = parts[1]  # берет вторую часть (время) 
    method = parts[2]  # берет третью часть (метод)
    endpoint = parts[3]  # берет четвертую часть (endpoint)
    status_code = parts[4]  # берет пятую часть (код статуса)
    time_to_answear = parts[5]  # берет шестую часть (время ответа)
    if status_code.startswith("5") or status_code.startswith("4"): 
        errors += 1  
print(f"Количество ошибок: {errors}")  

Количество строк: 50
Количество ошибок: 15


Следующим шаго будет определение самого частого endpoint, его можно реализовать разными путям, выбираем наиболее быстсрый, с таким же счетчиком, который будет считать был ли нужный показатель добавлен в словарь или нет, и в конце выделим max из этого словаря чтобы понять какой запрос был самым частым

In [4]:
counts = {}
errors = 0

with open("server_logs.txt", "r") as file:
    lines = file.readlines()
    print(f"Количество строк: {len(lines)}")
    
    for line in lines:
        parts = line.strip().split()
        date = parts[0]
        time = parts[1]
        method = parts[2]
        endpoint = parts[3]
        status_code = parts[4]
        time_to_answear = parts[5]
        
        if status_code.startswith("5") or status_code.startswith("4"):
            errors += 1
            
        if endpoint in counts:
            counts[endpoint] += 1
        else:
            counts[endpoint] = 1


max_endpoint = max(counts, key=counts.get)
print(f"Самый частый endpoint: {max_endpoint} — {counts[max_endpoint]} раз")
print(f"Количество ошибок: {errors}")

Количество строк: 50
Самый частый endpoint: /profile — 13 раз
Количество ошибок: 15


Таким образом мы видии, что самый частый endpoint: /profile. Он вызывался 13 раз. 

Для того чтобы посчитать среднеее время ответа сервера, мы пойем по простому пути, так как мы знаем общее число всех записей, их 50, мы можем посчитать при помощи среднего арифметического. Но время ответа записано в логах как 342ms, для начала мы преобразуем нужную нам часть в число и перместим ms, чтобы можноы бло работать с числами.

In [5]:
time_to_answear = int(parts[5].replace("ms", ""))

In [6]:
counts = {}
errors = 0

with open("server_logs.txt", "r") as file:
    lines = file.readlines()
    print(f"Количество строк: {len(lines)}")
    
    for line in lines:
        parts = line.strip().split()
        date = parts[0]
        time = parts[1]
        method = parts[2]
        endpoint = parts[3]
        status_code = parts[4]
        time_to_answear = int(parts[5].replace("ms", ""))
        
        if status_code.startswith("5") or status_code.startswith("4"):
            errors += 1
            
        if endpoint in counts:
            counts[endpoint] += 1
        else:
            counts[endpoint] = 1
        
    average_time = sum(int(line.strip().split()[5].replace("ms", "")) for line in lines) / len(lines)

max_endpoint = max(counts, key=counts.get)
print(f"Самый частый endpoint: {max_endpoint} — {counts[max_endpoint]} раз")
print(f"Количество ошибок: {errors}")
print(f"Среднее время ответа: {average_time:.2f} ms")


Количество строк: 50
Самый частый endpoint: /profile — 13 раз
Количество ошибок: 15
Среднее время ответа: 959.50 ms


Таким образом мы посчитали среднее время ответа сервера оно составляет 959.50 ms. 

Вывод: В данной работе был представлен анализ логов сервера. В ходе данной работы было выявлено, что: 
1. количество записей в логах равно 50
2. Самый частый endpoint: /profile, он вызывался 13 раз
3. Количество ошибок, который были с кодом 404 или 500 составило 15
4. Среднее время ответа сервера составила 959. 5 мс
Поэтим данным можно судить о работе сервера, его настройке, какие ошибки возникали у пользователей и т.д